secboot/keymgr/keymgr_luks2.go

```diff
@@ -0,0 +1,229 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package keymgr
+
+import (
+	"fmt"
+	"regexp"
+	"time"
+
+	sb "github.com/snapcore/secboot"
+
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/secboot"
+	"github.com/snapcore/snapd/secboot/keyring"
+	"github.com/snapcore/snapd/secboot/luks2"
+)
+
+const (
+	// key slot used by the encryption key
+	encryptionKeySlot = 0
+	// key slot used by the recovery key
+	recoveryKeySlot = 1
+	// temporary key slot used when changing the encryption key
+	tempKeySlot = recoveryKeySlot + 1
+)
+
+var (
+	sbGetDiskUnlockKeyFromKernel = sb.GetDiskUnlockKeyFromKernel
+	keyringAddKeyToUserKeyring   = keyring.AddKeyToUserKeyring
+)
+
+func getEncryptionKeyFromUserKeyring(dev string) ([]byte, error) {
+	const remove = false
+	const defaultPrefix = "ubuntu-fde"
+	// note this is the unlock key, which can be either the main key which
+	// was unsealed, or the recovery key, in which case some operations may
+	// not make sense
+	currKey, err := sbGetDiskUnlockKeyFromKernel(defaultPrefix, dev, remove)
+	if err != nil {
+		return nil, fmt.Errorf("cannot obtain current unlock key for %v: %v", dev, err)
+	}
+	return currKey, err
+}
+
+func isKeyslotNotActive(err error) bool {
+	match, _ := regexp.MatchString(`.*: Keyslot [0-9]+ is not active`, err.Error())
+	return match
+}
+
+func recoveryKDF() (*sb.KDFOptions, error) {
+	usableMem, err := osutil.TotalUsableMemory()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get usable memory for KDF parameters when adding the recovery key: %v", err)
+	}
+	// The KDF memory is heuristically calculated by taking the
+	// usable memory and subtracting hardcoded 384MB that is
+	// needed to keep the system working. Half of that is the mem
+	// we want to use for the KDF. Doing it this way avoids the expensive
+	// benchmark from cryptsetup. The recovery key is already 128bit
+	// strong so we don't need to be super precise here.
+	kdfMem := (int(usableMem) - 384*1024*1024) / 2
+	// max 1 GB
+	if kdfMem > 1024*1024*1024 {
+		kdfMem = (1024 * 1024 * 1024)
+	}
+	// min 32 KB
+	if kdfMem < 32*1024 {
+		kdfMem = 32 * 1024
+	}
+	return &sb.KDFOptions{
+		MemoryKiB:       kdfMem / 1024,
+		ForceIterations: 4,
+	}, nil
+}
+
+func sbKDFToLuksKDF(o *sb.KDFOptions) luks2.KDFOptions {
+	return luks2.KDFOptions{
+		TargetDuration:  o.TargetDuration,
+		MemoryKiB:       o.MemoryKiB,
+		ForceIterations: o.ForceIterations,
+		Parallel:        o.Parallel,
+	}
+}
+
+// AddRecoveryKeyToLUKSDevice adds a recovery key to a LUKS2 device. It the
+// devuce unlock key from the user keyring to authorize the change. The
+// recoveyry key is added to keyslot 1.
+func AddRecoveryKeyToLUKSDevice(dev string, recoveryKey secboot.RecoveryKey) error {
```

Comment:
I'm confused by this package, will it ultimately be changed to call the keymgr tool, or it exists mainly as helpers for the keymgr tool?

Reply:
I would keep the signature more like the original secboot ones with the device last perhaps?

Reply:
It's just helpers for the tool. The actual tool is at `cmd/snap-fde-keymgr`.